google/cloud/spanner/interval_test.cc

```diff
@@ -0,0 +1,205 @@
+
+  EXPECT_EQ(Interval(0, 2, -90, -absl::Hours(12)),
+            Interval(0, -1, -1, absl::Hours(12)));
+
+  // Check microsecond rounding.
+  EXPECT_EQ(Interval(absl::Microseconds(1.25)),
+            Interval(absl::Microseconds(1)));
+  EXPECT_EQ(Interval(-absl::Microseconds(1.25)),
+            Interval(-absl::Microseconds(1)));
+  EXPECT_EQ(Interval(absl::Microseconds(1.5)), Interval(absl::Microseconds(2)));
+  EXPECT_EQ(Interval(-absl::Microseconds(1.5)),
+            Interval(-absl::Microseconds(2)));
+
+  // Check that the logical value of an Interval used during comparison
+  // is able to represent values greater that (2^63-1) nanoseconds.
+  auto const max64 = std::numeric_limits<std::int64_t>::max();
+  EXPECT_GT(Interval(296, 6, 12), Interval(absl::Nanoseconds(max64)));
+}
+
+TEST(Interval, ArithmeticOperators) {
+  // Negation.
+  EXPECT_EQ(-Interval(), Interval());
+  EXPECT_EQ(-Interval(0, 10, 11, absl::Hours(12)),
+            Interval(0, -10, -11, -absl::Hours(12)));
+
+  // Addition/subtraction.
+  EXPECT_EQ(Interval(0, 1, 0) + Interval(0, 2, 0), Interval(0, 3, 0));
+  EXPECT_EQ(Interval(1, 0, 0, absl::Hours(24)) + Interval(1, 0, 3),
+            Interval(2, 0, 3, absl::Hours(24)));
+  EXPECT_EQ(Interval(1, 0, 1) - Interval(absl::Hours(24)),
+            Interval(1, 0, 1, -absl::Hours(24)));
+  EXPECT_EQ(Interval(0, 11, 0) + Interval(0, 1, 1), Interval(1, 0, 1));
+  EXPECT_EQ(Interval(absl::Hours(2) + absl::Minutes(50)) +
+                Interval(absl::Minutes(10)),
+            Interval(absl::Hours(3)));
+  EXPECT_EQ(Interval(absl::Hours(2) + absl::Minutes(50)) -
+                Interval(absl::Minutes(50)),
+            Interval(absl::Hours(2)));
+
+  // Multiplication/division.
+  EXPECT_EQ(Interval(0, 1, 0) * 2, Interval(0, 2, 0));
+  EXPECT_EQ(Interval(1, 6, 0) * 2, Interval(3, 0, 0));
+  EXPECT_EQ(Interval(absl::Hours(5) + absl::Minutes(5)) * 2,
+            Interval(absl::Hours(10) + absl::Minutes(10)));
+  EXPECT_EQ(Interval(0, 0, 15, absl::Hours(24)) * 3,
+            Interval(0, 1, 15, absl::Hours(72)));
+  EXPECT_EQ(Interval(0, 1, 15) * 3, Interval(0, 3, 45));
+  EXPECT_EQ(Interval(0, 1, 15) * 12, Interval(1, 0, 180));
+  EXPECT_EQ(Interval(0, 1, 0) / 30, Interval(0, 0, 1));
+  EXPECT_EQ(Interval(1, 0, 0) / 365, Interval(absl::Hours(24) * (360.0 / 365)));
+  EXPECT_EQ(Interval(1, 0, 0) / 12, Interval(0, 1, 0));
+  EXPECT_EQ(Interval(0, 1, 0) / 4, Interval(0, 0, 7, absl::Hours(12)));
+  EXPECT_EQ(Interval(0, 1, 0) * 0.5, Interval(0, 0, 15));
+  EXPECT_EQ(Interval(absl::Minutes(1)) * 0.5, Interval(absl::Seconds(30)));
+  EXPECT_EQ(Interval(0, 3, 30) / 30, Interval(0, 0, 4));
+  EXPECT_EQ(Interval(absl::Milliseconds(1001)) / 1'000'000 * 1'000,
+            Interval(absl::Microseconds(1001)));
+  EXPECT_EQ(Interval(absl::Minutes(1)) * 600, Interval(absl::Hours(10)));
+}
+
+TEST(Interval, Range) {
+  auto huge = Interval(178'000'000, 0, 0);
+  EXPECT_LT(-huge, huge);
+  EXPECT_EQ(std::string(huge), "178000000 years");
```

Comment:
I think these are the only tests for `Stringify()`. Consider adding some (more interesting) test cases, and also consider some tests for `operator<<` at the same time.

Reply:
The `RoundTrip` cases also cover `Stringify()` (which I renamed to `SerializeInterval()` to match the new `ParseInterval()`).  But I still added more cases there to make it clear that each branch is exercised.

`operator<<` is defined to use `operator std::string`, so it doesn't need extra functional tests, but I did add a new test just to verify that `<< Interval` works.